### PR TITLE
[UCC] Properly finalize unsuccessful collective posts

### PR DESCRIPTION
**Pull request by Aidyn-A**
This PR add a `ucc_collective_finalize` call if `ucc_collective_post` and `ucc_collective_triggered_post` were not successful. 
According to the [UCC documentation](https://openucx.github.io/ucc/api/v1.1/html/group___u_c_c___c_o_l_l_e_c_t_i_v_e_s.html):
```
On error, request handle becomes invalid, user is responsible to call ucc_collective_finalize to free allocated resources.
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

### CODE REVIEW

File: torch/csrc/distributed/c10d/UCCUtils.hpp [hunk 1]
```diff
@@ -31,6 +31,30 @@ namespace c10d {
...
+          logger->getLogPrefix(),                           \
+          _error_msg,                                       \
+          ", error code ",                                  \
+          result,                                           \
+          ": ",                                             \
+          ucc_status_string(result),                        \
+          ", system error code ",                           \
+          errno);                                           \
+      ucc_collective_finalize(_request);                    \
+      TORCH_CHECK(false, err);                              \
```
Comment (Aidyn-A) [inline torch/csrc/distributed/c10d/UCCUtils.hpp:54]: I do not think we need additional check for `ucc_collective_finalize` here.

**Comment by Aidyn-A** [comment]
@kwen2501 can you please take a look?

**Comment by kwen2501** [comment]
Cc @vtlam @kingchc , can you please take a look? Thanks!

File: torch/csrc/distributed/c10d/UCCUtils.hpp [hunk 2]
```diff
@@ -31,6 +31,33 @@ namespace c10d {
...
+          std::to_string(__LINE__),                         \
+          "] ",                                             \
+          logger->getLogPrefix(),                           \
+          _error_msg,                                       \
+          ", error code ",                                  \
+          result,                                           \
+          ": ",                                             \
+          ucc_status_string(result),                        \
+          ", system error code ",                           \
+          errno);                                           \
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/UCCUtils.hpp:52]: This error string preparation looks the same as the one above in `TORCH_UCC_CHECK`.
Do you mind consolidating them? Thanks!

File: torch/csrc/distributed/c10d/UCCUtils.hpp (hunk 1 above)
Comment (kwen2501) [reply torch/csrc/distributed/c10d/UCCUtils.hpp:57]: Would we lose the original error messages, i.e. `err` and `_error_msg`, if we `TORCH_UCC_CHECK`'ing `ucc_collective_finalize` here?

Comment (Aidyn-A) [reply torch/csrc/distributed/c10d/UCCUtils.hpp:57]: You are right, if `ucc_collective_finalize` is failed then the iriginal message `err` will be lost.

**Review by kwen2501 — APPROVED** [review]
Thanks for resolving the comments!
Looks good to me!

**Comment by Aidyn-A** [comment]
Thank you!

@pytorchbot merge

**Review by kingchc — COMMENTED** [review]
The PR looks good to me overall, I have two minor comments we might discuss/consider addressing here or in a separate PR. Thanks.

File: torch/csrc/distributed/c10d/UCCUtils.hpp [hunk 3]
```diff
@@ -8,27 +8,46 @@
...
+      _err = c10::str(                            \
+          "[",                                    \
+          std::string(__FILE__),                  \
+          ":",                                    \
+          std::to_string(__LINE__),               \
+          "] ",                                   \
+          logger->getLogPrefix(),                 \
+          _error_msg,                             \
+          ", error code ",                        \
+          result,                                 \
```
[the remark this answers is not in the document]
Comment (Aidyn-A) [reply torch/csrc/distributed/c10d/UCCUtils.hpp:23]: Good suggestion. Thanks!

**Pull request closed without merging**